# Algoritmo do Laboratório Nacional de Dados, visando fazer o parsing nos dados obtidos do DataSUS - Projeto Turing - Referente a Suícidios  
📄 Fonte dos dados (CSV):
https://s3.sa-east-1.amazonaws.com/ckan.saude.gov.br/SIM/DO23OPEN.csv

📘 Dicionário de Dados (Data Dictionary):
https://diaad.s3.sa-east-1.amazonaws.com/sim/Mortalidade_Geral+-+Estrutura.pdf

Site Base dos dados:
https://opendatasus.saude.gov.br/dataset/sim/

In [75]:
# Importando as bibliotecas necessárias
import pandas as pd

In [76]:
# Jogando o arquivo CSV para dentro do notebook
df = pd.read_csv(r"C:\Users\Kaue Mandarino\Downloads\DO23OPEN (1).csv") # Para o Python ler o endereço windows tem que por um "r" antes do caminho do arquivo

In [77]:
# Exibindo as primeiras 3 linhas do dataframe para ver se deu certo:
df

contador;"ORIGEM";"TIPOBITO";"DTOBITO";"HORAOBITO";"NATURAL";"CODMUNNATU";"DTNASC";"IDADE";"SEXO";"RACACOR";"ESTCIV";"ESC";"ESC2010";"SERIESCFAL";"OCUP";"CODMUNRES";"LOCOCOR";"CODESTAB";"CODMUNOCOR";"IDADEMAE";"ESCMAE";"ESCMAE2010";"SERIESCMAE";"OCUPMAE";"QTDFILVIVO";"QTDFILMORT";"GRAVIDEZ";"SEMAGESTAC";"GESTACAO";"PARTO";"OBITOPARTO";"PESO";"TPMORTEOCO";"OBITOGRAV";"OBITOPUERP";"ASSISTMED";"EXAME";"CIRURGIA";"NECROPSIA";"LINHAA";"LINHAB";"LINHAC";"LINHAD";"LINHAII";"CAUSABAS";"CB_PRE";"COMUNSVOIM";"DTATESTADO";"CIRCOBITO";"ACIDTRAB";"FONTE";"NUMEROLOTE";"DTINVESTIG";"DTCADASTRO";"ATESTANTE";"STCODIFICA";"CODIFICADO";"VERSAOSIST";"VERSAOSCB";"FONTEINV";"DTRECEBIM";"ATESTADO";"DTRECORIGA";"OPOR_DO";"CAUSAMAT";"ESCMAEAGR1";"ESCFALAGR1";"STDOEPIDEM";"STDONOVA";"DIFDATA";"NUDIASOBCO";"DTCADINV";"TPOBITOCOR";"DTCONINV";"FONTES";"TPRESGINFO";"TPNIVELINV";"DTCADINF";"MORTEPARTO";"DTCONCASO";"ALTCAUSA";"CAUSABAS_O";"TPPOS";"TP_ALTERA";"CB_ALT"
0        1;"1";"2";"14022023";"0930";"833";"330270";"10...                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
1        2;"1";"2";"14022023";"1730";"822";"221110";"12...                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
2        3;"1";"2";"14022023";"0500";"822";"220790";"10...                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
3        4;"1";"2";"14022023";"2235";"823";"231180";"15...                                                                                                                      

Não deu certo porque a função de ler CSV imagina sozinha que o separador seja uma virgula, e por isso não converte,  
 no caso em questão o separador do arquivo é um ';" como podemos.  
contador;"ORIGEM";"TIPOBITO";"DTOBITO";

In [78]:
# Explicitando o separador como ponto e vírgula para a leitura correta do arquivo
df = pd.read_csv(r"C:\Users\Kaue Mandarino\Downloads\DO23OPEN (1).csv", sep=';')

C:\Users\Kaue Mandarino\AppData\Local\Temp\ipykernel_38068\1699433278.py:2: DtypeWarning: Columns (65) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(r"C:\Users\Kaue Mandarino\Downloads\DO23OPEN (1).csv", sep=';')


Este warning: "DtypeWarning: Columns (65) have mixed types. Specify dtype option on import or set low_memory=False"  
quer dizer que o Python esta lendo coluna por coluna e atribuindo para cada uma um tipo, exemplo:  
Essa aqui é tipo numérico, essa é float, essa é texto...  
Mas para isso ele gasta memória, uma opção que consumiria menos processamento seria a gente olhar o arquivo original  
e especificar no parametro da funcão o tipo de cada coluna, opção mais viavél para arquivos mais pesados.

In [79]:
# Vamos exibir a cabeça do DataFrame para verificar se os dados foram carregados corretamente
df.head(3)

contador  ORIGEM  TIPOBITO   DTOBITO  HORAOBITO  NATURAL  CODMUNNATU  \
0         1       1         2  14022023      930.0    833.0    330270.0   
1         2       1         2  14022023     1730.0    822.0    221110.0   
2         3       1         2  14022023      500.0    822.0    220790.0   

       DTNASC  IDADE  SEXO  ...  TPRESGINFO  TPNIVELINV  DTCADINF  MORTEPARTO  \
0  10071954.0    468     1  ...         NaN         NaN       NaN         NaN   
1  12091956.0    466     1  ...         NaN         NaN       NaN         NaN   
2  10011953.0    470     2  ...         NaN         NaN       NaN         NaN   

   DTCONCASO  ALTCAUSA  CAUSABAS_O  TPPOS  TP_ALTERA  CB_ALT  
0        NaN       NaN         I10    NaN        NaN     NaN  
1        NaN       NaN        I678    NaN        NaN     NaN  
2        NaN       NaN        K750    NaN        NaN     NaN  

[3 rows x 86 columns]

Ok, uma vez carregado o dataframe vamos fazer o parsing (depara), tratando cada uma das 86 colunas numéricas para que fique em uma entrada mais palatável;  
Resumindo vamos desencripitar a Enigma rsrs.

In [80]:
# Removendo Coluna 'contador', uma vez que parece sujeira no DataFrame, creio que teria que ter usado o padrão reset_index() na importação  
# mas assim também funciona:
df = df.drop(columns=['contador'])
df.head(1)

ORIGEM  TIPOBITO   DTOBITO  HORAOBITO  NATURAL  CODMUNNATU      DTNASC  \
0       1         2  14022023      930.0    833.0    330270.0  10071954.0   

   IDADE  SEXO  RACACOR  ...  TPRESGINFO  TPNIVELINV  DTCADINF  MORTEPARTO  \
0    468     1      4.0  ...         NaN         NaN       NaN         NaN   

   DTCONCASO  ALTCAUSA  CAUSABAS_O  TPPOS  TP_ALTERA  CB_ALT  
0        NaN       NaN         I10    NaN        NaN     NaN  

[1 rows x 85 columns]

Agora vamos na sequencia direita para esquerda fazer o parser nas proximas uma a uma, são 86 o.o',  
A partir daqui precisamos estar com o dicionário em mãos:  
https://diaad.s3.sa-east-1.amazonaws.com/sim/Mortalidade_Geral+-+Estrutura.pdf

A primeira coluna é Origem:

Origem do registro. (1- Oracle; 2 - Banco estadual diponibilizado via FTP; 3 -
Banco SEADE; 9 - Ignorado)  
Indica o banco de dados origem, não vamos usar ela então:


In [81]:
df = df.drop(columns=['ORIGEM']) # Removendo a coluna 'ORIGEM' que não é necessária
df.head(1) # Exibindo a primeira linha do DataFrame para verificação

TIPOBITO   DTOBITO  HORAOBITO  NATURAL  CODMUNNATU      DTNASC  IDADE  \
0         2  14022023      930.0    833.0    330270.0  10071954.0    468   

   SEXO  RACACOR  ESTCIV  ...  TPRESGINFO  TPNIVELINV  DTCADINF  MORTEPARTO  \
0     1      4.0     1.0  ...         NaN         NaN       NaN         NaN   

   DTCONCASO  ALTCAUSA  CAUSABAS_O  TPPOS  TP_ALTERA  CB_ALT  
0        NaN       NaN         I10    NaN        NaN     NaN  

[1 rows x 84 columns]

A proxima coluna é TIPOOBITO:  
Tipo do óbito Óbito fetal: morte antes da expulsão ou da extração completa do   
corpo da Mãe, independentemente da duração da gravidez. Indica o óbito o  
fato de o feto, depois da expulsão do corpo materno, não respirar nem  
apresentar nenhum outro sinal de vida, como batimentos do coração,   
pulsações do cordão umbilical ou movimentos efetivos dos músculos de  
contração voluntária. (1-Fetal; 2-Não Fetal)  
Como trata de óbitos fetias não tem relação com nosso estudo (suicídio)  
Vamos tirar ela também:

Como pelo jeito vamos ficar repetindo esta remoção de colunas vou criar uma função para facilitar nossa vida:

In [82]:
def drop(col, df_name='df'):
    globals()[df_name] = globals()[df_name].drop(columns=[col])
    return globals()[df_name].head(1)

Agora ao inves de precisar digitar este codigo grande:  
df = df.drop(columns=['TIPOOBITO']) # Removendo a coluna 'ORIGEM' que não é necessária  
df.head(1) # Exibindo a primeira linha do DataFrame para verificação  
podemos digitar apenas:


In [83]:
drop('TIPOBITO')

DTOBITO  HORAOBITO  NATURAL  CODMUNNATU      DTNASC  IDADE  SEXO  RACACOR  \
0  14022023      930.0    833.0    330270.0  10071954.0    468     1      4.0   

   ESTCIV  ESC  ...  TPRESGINFO  TPNIVELINV  DTCADINF  MORTEPARTO  DTCONCASO  \
0     1.0  2.0  ...         NaN         NaN       NaN         NaN        NaN   

   ALTCAUSA  CAUSABAS_O  TPPOS  TP_ALTERA  CB_ALT  
0       NaN         I10    NaN        NaN     NaN  

[1 rows x 83 columns]

A próxima coluna é DTOBITO (data do óbito), só de olhar já da ver que ela indica ddmmaaa, então vamos transformar ela no formato padrão de data do python:

In [84]:
df['DTOBITO'] = pd.to_datetime(df['DTOBITO'], format='%d%m%Y')

ValueError: day is out of range for month, at position 35. You might want to try:
    - passing `format` if your strings have a consistent format;
    - passing `format='ISO8601'` if your strings are all ISO8601 but not necessarily in exactly the same format;
    - passing `format='mixed'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.

deu esse ValueError: day is out of range for month, at position 35.  
disse que a linha 35 esta com algum erro, vamos dar uma olhada nela:

In [85]:
df['DTOBITO'][35] # Exibindo a data de óbito da linha 35 para verificação

1072023

Ele deveria ser 01072023, mas perdeu o zero à esquerda.  
então antes de converter vamos preencher toda a coluna com 8 digitos  
utilizando a funcao zfill:

In [87]:
df['DTOBITO'] = (df['DTOBITO'].astype(str).str.zfill(8))
df['DTOBITO'][35] # Exibindo a data de óbito da linha 35 para verificação

'01072023'

In [88]:
# Uma vez tratados os zeros a esquerda, agora podemos converter para datetime
df['DTOBITO'] = pd.to_datetime(df['DTOBITO'], format='%d%m%Y')

Agora vamos ver se deu certo:  
Se retornar 0 → ✅ tudo certo  
Se > 0 → ⚠️ existem datas inválidas  

In [89]:
df['DTOBITO'].isna().sum()

0

🧠 O que significa isna()?  
isna() verifica se o valor está ausente / inválido.  
Em pandas, isso inclui:  
NaN → número ausente  
NaT → data inválida / ausente  
None → valor nulo  

In [ ]:
# Vamos ainda dar mais uma olhada na coluna:
df['DTOBITO']

0         2023-02-14
1         2023-02-14
2         2023-02-14
3         2023-02-14
4         2023-02-14
             ...    
1465605   2023-11-30
1465606   2023-12-12
1465607   2023-12-22
1465608   2023-12-22
1465609   2023-12-31
Name: DTOBITO, Length: 1465610, dtype: datetime64[ns]

deu certo

In [92]:
df.head(1)

DTOBITO  HORAOBITO  NATURAL  CODMUNNATU      DTNASC  IDADE  SEXO  \
0 2023-02-14      930.0    833.0    330270.0  10071954.0    468     1   

   RACACOR  ESTCIV  ESC  ...  TPRESGINFO  TPNIVELINV  DTCADINF  MORTEPARTO  \
0      4.0     1.0  2.0  ...         NaN         NaN       NaN         NaN   

   DTCONCASO  ALTCAUSA  CAUSABAS_O  TPPOS  TP_ALTERA  CB_ALT  
0        NaN       NaN         I10    NaN        NaN     NaN  

[1 rows x 83 columns]

A próxima coluna é HORAOBITO

In [94]:
df['HORAOBITO']

0           930.0
1          1730.0
2           500.0
3          2235.0
4           940.0
            ...  
1465605    2100.0
1465606    1220.0
1465607    1734.0
1465608    2030.0
1465609       NaN
Name: HORAOBITO, Length: 1465610, dtype: float64

Pelo jeito ela esta como decimal atoa, o ".0" (float) esta errado, ao que parece ela é inteiro e hhmm em um padrão de 24h  
vamos tratar

In [97]:
# Primeiro vamos remover o float:
df['HORAOBITO'] = df['HORAOBITO'].astype('int64')

IntCastingNaNError: Cannot convert non-finite values (NA or inf) to integer

Nao funcionou porque a coluna tem algum valor NA, vamos ver:

In [98]:
df['HORAOBITO'].isna().sum()

49218

Para tratar isso basta colocar o parametro "errors='coerce' que ele manterá os NA

In [ ]:
df['HORAOBITO'] = pd.to_datetime(df['HORAOBITO'], format='%H%M', errors='coerce').dt.time
df['HORAOBITO']

In [102]:
df.head(1)

DTOBITO HORAOBITO  NATURAL  CODMUNNATU      DTNASC  IDADE  SEXO  RACACOR  \
0 2023-02-14  09:30:00    833.0    330270.0  10071954.0    468     1      4.0   

   ESTCIV  ESC  ...  TPRESGINFO  TPNIVELINV  DTCADINF  MORTEPARTO  DTCONCASO  \
0     1.0  2.0  ...         NaN         NaN       NaN         NaN        NaN   

   ALTCAUSA  CAUSABAS_O  TPPOS  TP_ALTERA  CB_ALT  
0       NaN         I10    NaN        NaN     NaN  

[1 rows x 83 columns]

In [ ]:
# Parsing da coluna 'SEXO' para valores descritivos
df['SEXO'] = df['SEXO'].replace({
    1: 'Masculino',
    2: 'Feminino',
    0: 'Ignorado'
})

In [ ]:
df.head()  # Exibindo as primeiras linhas do DataFrame para verificação

ORIGEM    TIPOBITO   DTOBITO  HORAOBITO  NATURAL  CODMUNNATU      DTNASC  \
0       1  Fatalidade  14022023      930.0    833.0    330270.0  10071954.0   
1       1  Fatalidade  14022023     1730.0    822.0    221110.0  12091956.0   
2       1  Fatalidade  14022023      500.0    822.0    220790.0  10011953.0   
3       1  Fatalidade  14022023     2235.0    823.0    231180.0  15121943.0   
4       1  Fatalidade  14022023      940.0    841.0    411350.0   9041973.0   

   IDADE       SEXO  RACACOR  ...  TPRESGINFO  TPNIVELINV  DTCADINF  \
0    468  Masculino      4.0  ...         NaN         NaN       NaN   
1    466  Masculino      4.0  ...         NaN         NaN       NaN   
2    470   Feminino      NaN  ...         NaN         NaN       NaN   
3    479   Feminino      1.0  ...         NaN         NaN       NaN   
4    449   Feminino      4.0  ...         NaN           M       NaN   

   MORTEPARTO  DTCONCASO  ALTCAUSA  CAUSABAS_O  TPPOS  TP_ALTERA  CB_ALT  
0         NaN        NaN       NaN         I10    NaN        NaN     NaN  
1         NaN        NaN       NaN        I678    NaN        NaN     NaN  
2         NaN        NaN       NaN        K750    NaN        NaN     NaN  
3         NaN        NaN       NaN         I64      S        NaN     NaN  
4         NaN        NaN       NaN        C189      S        NaN     NaN  

[5 rows x 85 columns]

In [ ]:
# Convertendo a coluna 'DTOBITO' para o formato datetime
df['DTOBITO'] = pd.to_datetime(df['DTOBITO'], format='%d%m%Y')

ValueError: day is out of range for month, at position 35. You might want to try:
    - passing `format` if your strings have a consistent format;
    - passing `format='ISO8601'` if your strings are all ISO8601 but not necessarily in exactly the same format;
    - passing `format='mixed'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.

"ValueError: day is out of range for month, at position 35. You might want to try:"  
deu erro na linha 35, vamos ver o porque:

In [ ]:
df[35]

KeyError: 35

In [ ]:
df.to_csv(r"C:\... POR ENDEREÇO AQUI", sep=';', index=False)

PermissionError: [Errno 13] Permission denied: 'C:\\... POR ENDEREÇO AQUI'